orm/src/main/java/org/hibernate/search/v6poc/entity/orm/event/impl/FullTextIndexEventListener.java

```diff
@@ -196,19 +201,52 @@ private void processCollectionEvent(AbstractCollectionEvent event) {
 			//Should log really but we don't know if we're interested in this collection for indexing
 			return;
 		}
-		PersistentCollection persistentCollection = event.getCollection();
-		final String collectionRole;
-		if ( persistentCollection != null ) {
-			collectionRole = persistentCollection.getRole();
-		}
-		else {
-			collectionRole = null;
+
+		if ( isWorkable( context, entity ) ) {
+			ChangesetPojoWorker worker = context.getCurrentWorker( event.getSession() );
+			if ( dirtyCheckingEnabled ) {
+				PersistentCollection persistentCollection = event.getCollection();
+				String collectionRole = null;
+				if ( persistentCollection != null ) {
+					collectionRole = persistentCollection.getRole();
+				}
+				if ( collectionRole != null ) {
+					/*
+					 * Collection role will only be non-null for PostCollectionUpdateEvents.
+					 * For those events, we can pass the role to the worker
+					 * which can then decide whether to reindex based on whether the collection
+					 * has any impact on indexing.
+					 */
+					worker.update( event.getAffectedOwnerIdOrNull(), entity, collectionRole );
+				}
+				else {
+					/*
+					 * We don't know which collection is being changed,
+					 * so we have to default to reindexing, just in case.
+					 */
+					worker.update( event.getAffectedOwnerIdOrNull(), entity );
+				}
+			}
+			else {
+				worker.update( event.getAffectedOwnerIdOrNull(), entity );
+			}
 		}
+	}
 
-		if ( isWorkable( context, entity ) &&
-				( !dirtyCheckingEnabled || true ) ) { // TODO implement dirty checking based on the collection role
-			context.getCurrentWorker( event.getSession() )
-					.update( event.getAffectedOwnerIdOrNull(), entity );
+	public String[] getDirtyPropertyNames(PostUpdateEvent event) {
+		EntityPersister persister = event.getPersister();
+		final int[] dirtyProperties = event.getDirtyProperties();
+		if ( dirtyProperties != null && dirtyProperties.length > 0 ) {
+			String[] propertyNames = persister.getPropertyNames();
+			int length = dirtyProperties.length;
+			String[] dirtyPropertyNames = new String[length];
+			for ( int i = 0; i < length; i++ ) {
+				dirtyPropertyNames[i] = propertyNames[dirtyProperties[i]];
+			}
+			return dirtyPropertyNames;
```

Comment:
I find it a bit weird that you push null values for dirty properties.

Reply:
You're talking about the else branch below, I suppose?

You're right, I copy/pasted this from Search 5 without thinking twice.
I just checked, and `dirtyProperties` being null should not happen, ever, because of this in `org.hibernate.event.internal.DefaultFlushEntityEventListener#scheduleUpdate`:

```
		int[] dirtyProperties = event.getDirtyProperties();
		if ( event.isDirtyCheckPossible() && dirtyProperties == null ) {
			if ( !intercepted && !event.hasDirtyCollection() ) {
				throw new AssertionFailure( "dirty, but no dirty properties" );
			}
			dirtyProperties = ArrayHelper.EMPTY_INT_ARRAY;
		}
```

It might be empty, though.

I'll add a commit so that we return an empty array instead, just to avoid any NPE nightmare in the future.